### PR TITLE
[WIP] Add an Og::createMembership helper

### PR DESCRIPTION
**Pull request by damiankloip**
I think we need one of these. I still need to add tests.

### CODE REVIEW

**Comment by amitaibu** [comment]
What's needed here is port of og_group ()

**Comment by damiankloip** [comment]
ok, can we work with what we have here and add stuff from og_group? I think passing in entities as params is better. Also, `og_group` doesn't really work as a name anymore so much?

**Comment by damiankloip** [comment]
Hang on, shouldn't this just start simple and be replacing `og_membership_create`? The og_group function has a whole bunch of other dependencies we don't have yet I think.

**Comment by amitaibu** [comment]
> shouldn't this just start simple and be replacing og_membership_create?

Ok, fair enough

File: src/Og.php [hunk 1]
```diff
@@ -419,4 +421,27 @@ public static function getSelectionHandler($entity, $bundle, $field_name, array
     return \Drupal::service('plugin.manager.entity_reference_selection')->createInstance('og:default', $options);
   }
 
+  /**
+   * Creates and saves an OG membership.
```
Comment (amitaibu) [inline src/Og.php:451]: this should follow the strictness of `og_membership_create()`:
- validate the group audience field
- Throw exception if membership already exists.

Comment (damiankloip) [inline src/Og.php:451]: Roger that.

Comment (damiankloip) [inline src/Og.php:451]: Currently it doesn't do the second one, should we add that in here?

Comment (damiankloip) [inline src/Og.php:451]: I think also I should remove saving from there.

**Comment by damiankloip** [comment]
Hmm, that is not right. I don't think we can use this function directly like that, we need `og_group` too. As create membership relies on having the field name existing on the entity etc...

**Comment by damiankloip** [comment]
Some other dependencies that would be good to resolve before unit tests, like https://github.com/amitaibu/og/pull/64

File: src/Og.php (hunk 1 above)
Comment (amitaibu) [inline src/Og.php:451]: > Currently it doesn't do the second one, should we add that in here?

If I remember correctly, if the membership already exists then we will have a mysql error, due to duplicate primary keys. I prefer to catch this error and throw a more descriptive error for a better DX.

File: src/Og.php [hunk 2]
```diff
@@ -446,4 +449,36 @@ public static function getSelectionHandler($entity, $bundle, $field_name, array
...
+   * @param \Drupal\Core\Entity\EntityInterface $group
+   * @param string $membership_type
+   * @param string $field_name
+   *
+   * @return \Drupal\og\Entity\OgMembership
+   */
+  public static function createMembership(FieldableEntityInterface $entity, EntityInterface $group, $membership_type = OG_MEMBERSHIP_TYPE_DEFAULT, $field_name = OG_AUDIENCE_FIELD) {
+    $field_definition = $entity->getFieldDefinition($field_name);
+
+    if (!$field_definition) {
```
Comment (amitaibu) [inline src/Og.php:463]: I think that maybe we should also have these two checks in `Drupal\og\Entity\OgMembership::preSave` as well

**Comment by amitaibu** [comment]
> Hmm, that is not right. I don't think we can use this function directly like that, we need og_group too

Sorry, not sure I follow what you meant here.

**Comment by damiankloip** [comment]
The Field name beign checked like this bu default causes problems, e.g. you try to use this in hook_entity_insert to create a membership for the owner of the group. So user is passed in, but default value for $field_name is OG_AUDIENCE_FIELD, so it fails as user does not have one of those. So should we change the default value to NULL? or require cases like that pass NULL themselves, then in this method, we check `isset($field_name) && $field_definition` type thing.

**Comment by damiankloip** [comment]
At the moment everything is being create with 'og_group_ref' as the field. So I think we need to fix that anyway?

**Comment by amitaibu** [comment]
@damiankloip indeed, that's where `Og::group()` will come - it will create the membership under the matching field (as in D7's `og_group`).

This `Og::createMembership` is a low level API, that doesn't concern itself with which audience field to use - it's using the one passed to it.

**Comment by amitaibu** [comment]
So yeah, `$field_name` can be a required field, as it will always be populated by a higher level method.

**Comment by damiankloip** [comment]
@amitaibu ok, I don't understand fully how this will work with things like adding the group owner. If we are validating the field, that field will not exist on a user so we will always get an error? In D7 were user memberships like this just saved with the default field? if so, we will need to account for this somehow here I think.

**Comment by amitaibu** [comment]
> If we are validating the field, that field will not exist on a user so we will always get an error?

An audience field (not necessarily the default one) _must_ exist on the user. In D7 when you created a group, we were kind enough to automatically add an audience field on the user for you, in case it didn't already exist.

**Comment by amitaibu** [comment]
In other words - the only way a group content (e.g. user, node-article) may belong to a group - is by having an audience field.

**Comment by damiankloip** [comment]
Oh, so implicitly added to the user. So all group owners would have this reference visible on a user audience field? I guess that's why there is the special user targets permission override stuff in some access checking?

**Comment by amitaibu** [comment]
> Oh, so implicitly added to the user.

If an existing field isn't there- yes. Otherwise, for example the "subscribe to group" won't appear.

>  I guess that's why there is the special user targets permission override stuff in some access checking?

Yes, only site admins with `administer group` by default have `edit` access to that field.

Also that field is used by `og_register` module to  allow register + auto-subscribe to a group.

**Comment by amitaibu** [comment]
Needs a re-roll, and some minor comments. Test can be in follow up :)

**Comment by damiankloip** [comment]
Thanks for the comments @amitaibu, I think we need to consider how we want to move forward with the user group ref field, as we need it here for this to work anyway. I think we should consider adding this to users on install. Thoughts? That seems better than always checking for the field just so we can add it the first time we need it.

**Comment by amitaibu** [comment]
> I think we should consider adding this to users on install. Thoughts?

We can't, because the field on the user references a certain entity type, or bundles of that entity type. There is no guarantee that for example we will have groups of type node.

I think that at this point we should continue with the D7 logic - and automatically add a field to the user upon `createField` (if none already exists yet).

Anyway, bare in mind that the membership you are working on here, isn't limited to users. The relation of a group content (node) to another group (node) is also via membership.

In short, I believe the PR is on its right track, and I think once we reach `og_group()` porting, those pieces will fall into place. Sounds ok? :)

**Comment by damiankloip** [comment]
I am good with that, and yes og_group and the unsubscribe version we need.

I get that the membership can be between anything, the problem we have here is that because we now validate the field in `createMembership` you cannot add anything if the user doesn't have that field already. Which we can obviously get working on an actual install, but don't have in tests.

**Comment by amitaibu** [comment]
We should be able to pick up this after #261 gets in. It would be a nice addition.

**Comment by amitaibu** [comment]
I will close this and re-open a new branch, branching out of #261
